### PR TITLE
Google Analytics Tracking

### LINKED ISSUE #119: Google Analytics
### Metrics to track

**Message Sarah for login credentials.**

**Please include device type as a parameter in all requests.**

| GA Event Name | GA Event Properties |
|-|-|
| User clicked a state in the state list | Name of state clicked | 
| User clicked a state on the map | Name of state clicked |
| User opened Glossary |
| User opened Scoring Guide |
| User opened Methodology |
| User opened Partner with FreeFrom form |
| User submitted Partner with FreeFrom form |
| User opened Build Collective Survivor Power form |
| User submitted Build Collective Survivor Power form |
| User opened Share Your Policy Ideas form |
| User submitted Share Your Policy Ideas form |
| User opened Give Feedback form |
| User submitted Give Feedback form |
| User opened Subscribe to State Updates form | Current page (home or state) |
| User submitted Subscribe to State Updates form | Current page (home or state) |
| User opened Report Missing/Outdated Information form | Current page (home or state) |
| User submitted Report Missing/Outdated Information form | Current page (home or state) |
| User clicked "Share" button from homepage | Current page (home or state) |
| User clicked "Share" button from state page | Current page (home or state) |

### PR DESCRIPTION
**Pull request by egiurleo**
Closes #119.

This PR adds the rest of Google Analytics tracking to the policy map. Here's an explanation of how all these actions get recorded:

Action | How it's referenced in GA
|-|-|
User clicked a state in the state list | Event with category `Click`, action `State list`, and the name of the state as the label
User clicked a state on the map | Event with category `Click`, action `Map`, and the name of the state as the label
User opened Glossary |  Modal view with `Glossary` as the name
User opened Scoring Guide |  Modal view with `Scoring Guide` as the name
User opened Methodology |  Page view with methodology URL
User opened Partner with FreeFrom form |  Page view with partner with FreeFrom URL
User submitted Partner with FreeFrom form |  Event with category `Form`, action `Submission`, and `partner-with-freefrom` as the label
User opened Build Collective Survivor Power form |  Page view with build collective survivor power URL
User submitted Build Collective Survivor Power form |  Event with category `Form`, action `Submission`, and `build-survivor-power` as the label
User opened Share Your Policy Ideas form |  Page view with share your policy ideas URL
User submitted Share Your Policy Ideas form |  Event with category `Form`, action `Submission`, and `share-your-policy-ideas` as the label
User opened Give Feedback form |  Page view with give feedback URL
User submitted Give Feedback form |  Event with category `Form`, action `Submission`, and `give-feedback` as the label
User opened Subscribe to State Updates form |  Modal view with `Sign Up for State Updates` as the name
User submitted Subscribe to State Updates form | Event with category `Form`, action `Submission`, and `subscribe-to-state-updates` as the label
User opened Report Missing/Outdated Information form | Modal view with `Report Missing Info` as the name
User submitted Report Missing/Outdated Information form | Event with category `Form`, action `Submission`, and `report-missing-info` as the label
User clicked "Share" button from homepage | Event with category `Click`, action `Share on Twitter` or `Share on Facebook`, and `homepage` as the label
User clicked "Share" button from state page | Event with category `Click`, action `Share on Twitter` or `Share on Facebook`, and `state page` as the label

### CODE REVIEW

File: components/StatesList.js [hunk 1]
```diff
@@ -20,7 +21,10 @@ function StatesList({ states }) {
                     ? <ul style={{backgroundColor: 'white', padding: 7}}>
                         {visibleStates.map(state => (
                             <li key={state.code}>
-                                <StateCard state={state} />
+                                <StateCard
+                                    state={state}
+                                    onLearnMoreClick={() => { trackEvent({ category: 'Click', action: 'State list', label: state.name}) }}
```
Comment (egiurleo) [inline components/StatesList.js:26]: This gets passed down to the `Learn More` button in the state list and records a click on a specific state card in the list.

File: components/Tracking.js [hunk 2]
```diff
@@ -1,20 +1,10 @@
 import React, { useEffect } from "react";
 import ReactGA from "react-ga"
 import PropTypes from "prop-types"
-import { getCookiesFromLocalStorage } from 'utils';
+import { trackPageView } from 'utils';
```
Comment (egiurleo) [inline components/Tracking.js:4]: I extracted this into a utility method so we could hypothetically reuse it if we needed it.

File: components/UsMap.js [hunk 3]
```diff
@@ -164,6 +164,7 @@
                 // Navigate to state on click
                 const { name } = d.properties
                 window.location.href = `${window.location.href}states/${toSlug(name)}`
+                trackEvent({category: 'Click', action: 'Map', label: name})
```
Comment (egiurleo) [inline components/UsMap.js:167]: This records a click on a state in the map.

File: components/common/ShareButtons.js [hunk 4]
```diff
@@ -2,14 +2,16 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
 import { faFacebook, faTwitter } from '@fortawesome/free-brands-svg-icons';
+import { trackEvent } from 'utils';
 
-const ShareButtons = ({ className }) => (
+const ShareButtons = ({ className, context }) => (
```
Comment (egiurleo) [inline components/common/ShareButtons.js:7]: We use the `context` prop to determine if this is being rendered on the home page or state page, and then use that to record a click on either of the buttons.

File: components/common/StateUpdates.js [hunk 5]
```diff
@@ -45,7 +45,11 @@ const StateUpdates = () => {
                     method="post"
                     name="mc-embedded-subscribe-form"
                     noValidate
-                    onSubmit={() => setSubmitted(true)}
+                    onSubmit={() => {
+                        setSubmitted(true)
+                        trackEvent({ category: 'Form', action: 'Submission', label: 'subscribe-state-updates'})
```
Comment (egiurleo) [inline components/common/StateUpdates.js:50]: Record a submission of the "Subscribe to state updates" form.

File: components/modal/Modal.js [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
+import { trackModal } from "utils";
 
 export default function Modal({ className, text, target, title, children }) {
     return (
         <div>
-            <ModalButton text={ text } target={ target } />
+            <ModalButton
+                text={ text }
+                target={ target }
+                onClick={ () => { trackModal(title) } }
```
Comment (egiurleo) [inline components/modal/Modal.js:14]: Track modal opening, including the title of the modal.

File: pages/_app.js [hunk 7]
```diff
@@ -11,6 +12,10 @@ export default function MyApp({ Component, pageProps }) {
             window.$ = window.jQuery = $;
             return import("bootstrap");
         });
+
+        if(getCookiesFromLocalStorage && process.env.NEXT_PUBLIC_GA_ID) {
+            ReactGA.initialize(process.env.NEXT_PUBLIC_GA_ID);
+        }
```
Comment (egiurleo) [inline pages/_app.js:18]: Initializing Google Analytics for use throughout the app.

File: utils/forms.js [hunk 8]
```diff
@@ -53,6 +54,7 @@ export const submitForm = path => {
                 throw new Error(message)
             }
             setStatus({success: true})
+            trackEvent({ category: 'Form', action: 'Submission', label: path })
```
Comment (egiurleo) [inline utils/forms.js:57]: Track a form submission, including the name of the form we're submitting using the `path` argument.

File: utils/tracking.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import ReactGA from "react-ga"
+import { getCookiesFromLocalStorage } from 'utils';
+
+export const trackPageView = (url) => {
+    if(getCookiesFromLocalStorage && process.env.NEXT_PUBLIC_GA_ID) {
```
Comment (egiurleo) [inline utils/tracking.js:5]: We only want to send info to GA if the user has agreed to cookies and we're in an environment with the `NEXT_PUBLIC_GA_ID` variable.

**Review by sarahgreen — CHANGES_REQUESTED** [review]
@egiurleo Hm, I don't think the specific events are populating in GA. This is what the 'real time' tab looks like after clicking around a bunch.
Screen Shot 2021-05-05 at 8 46 05 PM

Worst case scenario, if we can't figure out how to pass through the events, we can still merge this in since we're capturing the number of overall visitors / active users.

**Comment by egiurleo** [comment]
@sarahgreen this hasn't been deployed yet, right? So how could it be populating? (If I remember correctly our other change only started working once it was on staging, right?)

**Comment by sarahgreen** [comment]
@egiurleo Ahh you're totally right. I got confused because of the active users showing up in GA, but that's because of the change we already merged into staging. Sorry! 😅

**Review by sarahgreen — APPROVED** [review]

**Comment by egiurleo** [comment]
@sarahgreen All good :) Thanks for being proactive and making sure everything works.

**Pull request merged**